finagle-redis/src/main/scala/com/twitter/finagle/redis/dispatch/SubscribeDispatcher.scala

```diff
@@ -0,0 +1,38 @@
+package com.twitter.finagle.redis.dispatch
+
+import com.twitter.finagle.dispatch.GenSerialClientDispatcher
+import com.twitter.finagle.redis.protocol._
+import com.twitter.finagle.transport.Transport
+import com.twitter.util.{Future, Promise}
+import java.util.concurrent.atomic.AtomicReference
+
+class SubscribeDispatcher(trans: Transport[SubscribeCommand, Reply])
+    extends GenSerialClientDispatcher[SubscribeCommand, Unit, SubscribeCommand, Reply](trans) {
+
+  private var listener = new AtomicReference[SubscribeListener]
+
+  loop()
+
+  private[this] def handleMessage(reply: Reply) {
+    listener.get().onMessage(reply)
+    loop()
+
+}
+
+  private[this] def loop(): Unit =
+    trans.read().onSuccess { reply =>
+      listener.get().onMessage(reply)
+      loop()
+    }
+    .onFailure { ex =>
+      Option(listener.get()).map(_.onException(ex))
+    }
+
+  protected def dispatch(req: SubscribeCommand, p: Promise[Unit]): Future[Unit] =
+    trans.write(req).onSuccess { _ => p.setDone() }
+
+  override def apply(req: SubscribeCommand): Future[Unit] = {
+    listener.compareAndSet(null, req.listener)
```

Comment:
All commands created by the same SubscribeClient instance carries the same listener, which is the SubscribeClient instance itself. It would be better if the SubscribeClient can be passed as a constructor argument for the dispatcher.